compiler/sem/semexprs.nim

```diff
@@ -366,29 +368,30 @@ proc semConv(c: PContext, n: PNode): PNode =
                                              inputVal: op,
                                              targetTyp: result.typ))
 
-proc semCast(c: PContext, n: PNode): PNode =
+proc semCast(c: PContext, n: PNode): CheckedAst =
   ## Semantically analyze a casting ("cast[type](param)")
   checkSonsLen(n, 2, c.config)
   let
-    targetType = semTypeNode(c, n[0], nil)
+    typeExpr   = semTypeNode2(c, n[0], nil)
+    targetType = typeExpr.typ
     castedExpr = semExprWithType(c, n[1])
   
   if tfHasMeta in targetType.flags:
-    result = c.config.newError(n,
+    result = c.config.newError2(n,
```

Comment:
Looking at this and the example below, maybe a proc like:

`proc initWith(x: var CheckedAst, n: PNode, diag: PAstDiag = nil)`

If `diag` is `nil` then `hasError` would be `false`. Then maybe something like this would be nice:

```nim
func isError*(x: CheckedAst): bool {.inlne.} = x.n != nil and x.n.kind == nkError
func isInitialized*(x: CheckedAst): bool {.inline.} = x.n != nil

template initIfNotError*(x: var CheckedAst, n: PNode, body: typed) =
  assert x.isError or not x.isInitialized
  if not x.isError:
    x.initWith(n)
    body

# also/or this would allow a `nil` `diag` to keep on the happy path
template initIfNotError*(x: var CheckedAst, n: PNode, diag: PAstDiag, body: typed) =
  assert x.isError or not x.isInitialized
  if diag != nil:
    x.initAsError(n, diag)
  if not x.isError:
    x.initWith(n)
    body
```

which would allow things like:

```nim
  let
    typeExpr   = semTypeNode2(c, n[0], nil)
    targetType = typeExpr.typ
    castedExpr = semExprWithType(c, n[1])
    diag =
      if tfHasMeta in targetType.flags:
        PAstDiag(kind: adSemCannotCastToNonConcrete, wrongType: targeType)
      elif not isCastable(c, targetType, castedExpr.typ):
        PAstDiag(kind: adSemCannotCastTypes, ...)

  result.initIfNotError(n, diag):
    result.typ = targeType
    ...
```

Reply:
That does collapse the code quite a bit, I like it. Each template parameter appearing in the body multiple times does seem like it could easily become a problem, however.